src/wallet/wallet.cpp

```diff
@@ -2857,204 +2846,169 @@ bool CWallet::CreateTransactionInternal(
-                    setCoins.clear();
-                    int change_spend_size = CalculateMaximumSignedInputSize(change_prototype_txout, this);
-                    // If the wallet doesn't know how to sign change output, assume p2sh-p2wpkh
-                    // as lower-bound to allow BnB to do it's thing
-                    if (change_spend_size == -1) {
-                        coin_selection_params.change_spend_size = DUMMY_NESTED_P2WPKH_INPUT_SIZE;
-                    } else {
-                        coin_selection_params.change_spend_size = (size_t)change_spend_size;
-                    }
-                    if (!SelectCoins(vAvailableCoins, nValueToSelect, setCoins, nValueIn, coin_control, coin_selection_params, bnb_used))
-                    {
-                        // If BnB was used, it was the first pass. No longer the first pass and continue loop with knapsack.
-                        if (bnb_used) {
-                            coin_selection_params.use_bnb = false;
-                            continue;
-                        }
-                        else {
-                            error = _("Insufficient funds");
-                            return false;
-                        }
-                    }
-                } else {
-                    bnb_used = false;
-                }
+            CAmount selected_eff = 0; // selected effective value
+            for (const CInputCoin& coin : setCoins) {
+                selected_eff += coin.effective_value;
+            }
 
-                const CAmount nChange = nValueIn - nValueToSelect;
-                if (nChange > 0)
+            const CAmount nChange = input_sum - nValue;
+            if (nChange > 0)
+            {
+                // Fill a vout to ourself
+                CTxOut newTxOut(nChange, scriptChange);
+
+                // Cost of change is the cost of creating the change output + cost of spending the change output in the future.
+                // For creating the change output now, we use the effective feerate.
+                // For spending the change output in the future, we use the discard feerate for now.
+                // So cost of change = (change output size * effective feerate) + (size of spending change output * discard feerate)
+                CAmount change_fee = coin_selection_params.m_effective_feerate.GetFee(coin_selection_params.change_output_size);
+                CAmount cost_of_change = coin_selection_params.m_discard_feerate.GetFee(coin_selection_params.change_spend_size) + change_fee;
+
+                // Never create dust outputs; if we would, just
+                // add the dust to the fee.
+                // When the selected_eff is within the exact match range
+                // (nValue + not_input_fees + cost_of_change), don't make change.
+                if (IsDust(newTxOut, coin_selection_params.m_discard_feerate) || selected_eff <= nValue + not_input_fees + cost_of_change)
```

Comment:
This is partly to confirm my understanding, partly a suggestion for clarity:
```suggestion
                // There are two cases in which we drop the change to fees:
                // 1. The change output would be dust.
                // 2. We found an "exact match," i.e. the difference between the total effective value of the selected
                // coins and nTargetValue is less than or equal to the cost of the change output.
                if (IsDust(newTxOut, coin_selection_params.m_discard_feerate) ||
                    selected_eff - (nValue + not_input_fees) <= cost_of_change)
```

Reply:
Yes, that is correct. I've changed the comment to be similar.